src/server/pkg/work/work.go

```diff
@@ -119,50 +119,49 @@ type ProcessFunc func(context.Context, *Task, *Task) error
 type Worker struct {
 	etcdClient          *etcd.Client
 	taskCol, subtaskCol col.Collection
-	processFunc         ProcessFunc
 }
 
 // NewWorker creates a new worker.
-func NewWorker(etcdClient *etcd.Client, etcdPrefix string, processFunc ProcessFunc) *Worker {
+func NewWorker(etcdClient *etcd.Client, etcdPrefix string, taskNamespace string) *Worker {
 	return &Worker{
-		etcdClient:  etcdClient,
-		taskCol:     collection(etcdClient, path.Join(etcdPrefix, taskPrefix)),
-		subtaskCol:  collection(etcdClient, path.Join(etcdPrefix, subtaskPrefix)),
-		processFunc: processFunc,
+		etcdClient: etcdClient,
+		taskCol:    collection(etcdClient, path.Join(etcdPrefix, taskPrefix, taskNamespace)),
+		subtaskCol: collection(etcdClient, path.Join(etcdPrefix, subtaskPrefix, taskNamespace)),
 	}
 }
 
 // Run runs the worker with the given context.
 // The worker will continue to watch the task collection until the context is cancelled.
-func (w *Worker) Run(ctx context.Context) error {
+func (w *Worker) Run(ctx context.Context, processFunc ProcessFunc) error {
 	return w.taskCol.ReadOnly(ctx).WatchF(func(e *watch.Event) (retErr error) {
 		var id string
 		taskInfo := &TaskInfo{}
 		if err := e.Unmarshal(&id, taskInfo); err != nil {
 			return err
 		}
 		if taskInfo.State == State_RUNNING {
-			return w.processTask(ctx, taskInfo.Task)
+			if err := w.processTask(ctx, taskInfo.Task, processFunc); err != nil && err.Error() != "context canceled" {
+				return err
+			}
 		}
 		return nil
 	})
 }
 
-func (w *Worker) processTask(ctx context.Context, task *Task) error {
-	var eg errgroup.Group
+func (w *Worker) processTask(ctx context.Context, task *Task, processFunc ProcessFunc) error {
 	ctx, cancel := context.WithCancel(ctx)
+	eg, ctx := errgroup.WithContext(ctx)
```

Comment:
several changes to do with canceling in this function - I'm not entirely sure this is right, but it works more than it did before

Reply:
Yeah, this is one of those things I was kicking down the road (you can see a note I made about error propagation further up). I do agree with the new context level for handling errors, but I think you should instead think of it as a context level for task completion (success/killed/failure) and a context level for errors (something in one of the goroutines errored). Therefore, if `Wait` returns an non-nil error, you would return nil if the context for task completion was cancelled and the error otherwise. This would mean that you would not have to do that weird string check for context cancelled in the calling function. This does not really solve error propagation though because it is not getting communicated back to the master (or retried by the worker), so there is still more work, but I guess you were hitting the classic stuck in an error group because a goroutine wasn't returning.